packages/keyring-controller/src/KeyringController.ts

```diff
@@ -1228,40 +1717,336 @@ export class KeyringController extends BaseController<
   }
 
   /**
-   * Sync controller state with current keyring store
-   * and memStore states.
+   * Create new vault And with a specific keyring
    *
-   * @fires KeyringController:stateChange
+   * Destroys any old encrypted storage,
+   * creates a new encrypted store with the given password,
+   * creates a new wallet with 1 account.
+   *
+   * @fires KeyringController:unlock
+   * @param password - The password to encrypt the vault with.
+   * @param keyring - A object containing the params to instantiate a new keyring.
+   * @param keyring.type - The keyring type.
+   * @param keyring.opts - Optional parameters required to instantiate the keyring.
+   * @returns A promise that resolves to the state.
    */
-  #fullUpdate() {
-    const { vault } = this.#keyring.store.getState();
-    const { keyrings, isUnlocked, encryptionKey, encryptionSalt } =
-      this.#keyring.memStore.getState();
-
-    this.update(() => ({
-      vault,
-      keyrings,
-      isUnlocked,
-      encryptionKey,
-      encryptionSalt,
-    }));
+  async #createNewVaultWithKeyring(
+    password: string,
+    keyring: {
+      type: string;
+      opts?: unknown;
+    },
+  ): Promise<KeyringControllerMemState> {
+    if (typeof password !== 'string') {
+      throw new TypeError(KeyringControllerError.WrongPasswordType);
+    }
+    this.#password = password;
+
+    await this.#clearKeyrings();
+    await this.#createKeyring(keyring.type, keyring.opts);
+    this.#setUnlocked();
+    return this.#getMemState();
+  }
+
+  /**
+   * Update the controller state with its current keyrings.
+   */
+  async #updateKeyringsInState(): Promise<void> {
+    const keyrings = await Promise.all(this.#keyrings.map(displayForKeyring));
+    this.update((state) => {
+      state.keyrings = keyrings;
+    });
   }
```

Comment:
This is the equivalent of `EthKeyringController.#updateMemStoreKeyrings`: https://github.com/MetaMask/KeyringController/blob/0134de248cc16126b9dcb880a4d96776f8dbc043/src/KeyringController.ts#L358

Reply:
Perhaps `#updateStateKeyrings` would reflect the old name better, but I'm not sure what would be the best naming for this

```suggestion
  /**
   * Update the controller state with its current keyrings.
   */
  async #updateStateKeyrings(): Promise<void> {
    const keyrings = await Promise.all(this.#keyrings.map(displayForKeyring));
    this.update((state) => {
      state.keyrings = keyrings;
    });
  }
```

Reply:
I like `updateKeyringsInState` over `updateStateKeyrings` as it reads better. However it might be nice to convey that it's not taking a set of keyrings but rather it's persisting the keyrings that have already been saved. So perhaps `persistKeyrings`? I don't know if we already use "persist" elsewhere to mean "write to state", but it seems to fit.

Reply:
Unfortunately we already have a method called `persistAllKeyrings`, which saves the contents of the `#keyrings` array to the encrypted (and persisted) storage - the vault

This function instead, syncs the Controller state with the contents of the `#keyrings` array, so we are not persisting keyrings anywhere, but we are updating their representation in the controller state

Reply:
Ah, gotcha. Okay, `updateKeyringsInState` seems fine then.